chore: document maintenance policy (#55)
aws/jsii-compiler: SUPPORT.md

   keep the necessary effort low.

1. **End-of-Support**: Release lines in **End-of-Support** are considered
   defunct and are no longer maintained. They no longer receive new features,
   bug fixes, or security updates. **End-of-Support** releases may continue to
   be usable for a period of time, but they could be indirectly broken by
   changes in newer releases at any time. Users are advised to migrate away from
   **End-of-Support** release lines at the earliest convenience.

### Stage Transitions

```mermaid
---
title: Release Line Lifecycle
---
stateDiagram-v2
  direction LR

  EoL: End-of-Support

  [*] --> Current
  Current --> Maintenance : new release line started
  Maintenance --> EoL : after 6 months
  EoL --> [*]
```

Whenever a new release line is started (typically with a new `x.y.0` release,
excluding pre-releases), it becomes **Current** and the release line it replaced
immediately enters the **Maintenace** stage.

Releases stay in the **Maintenance** stage for a minimum of 6 months before they
reach **End-of-Support**.

Once a release line has reached **End-of-Support**, it remains in this stage
indefinitely.

### Timelines & Communication

> The `typescript` compiler mainteners start a new release line on a quarterly

Reviewer: ```suggestion
The `typescript` compiler maintainers start a new release line on a quarterly
```

> basis, and users should expect the `jsii` compiler does the same. Users should

Reviewer: ```suggestion
basis, and users should expect the `jsii` compiler maintainers to do the same. Users should
```

expect some small code changes may be necessary when upgrading from one release
line to another, as a result of breaking changes introduced by the `typescript`
compiler or `jsii` itself. In either case, the release notes for the first

> release in a line includes all relevant information about the breaking changes

Reviewer: ```suggestion
release in a line includes the relevant information about the breaking changes
```

Reviewer: Suggestions says the same, but it's less absolute, also a non-blocker.

and instructions for users to fix their codebase.

The `jsii` compiler has a built-in mechanism to inform users that they are using
an unsupported or soon-to-become unsupported release line. A message is emitted
to `STDERR` when invoking the `jsii` command if the current release is:

- **End-of-Support** already

> - Planned to move to **End-of-Support** within the next 30 calendar days

Author: This technically doesn't exist yet, but I have a plan that is ready to execute on.

The built-in messaging can only be expected to deliver the message if the
compiler is on the latest available release within the line, or if the
environment in which the compiler is run is able to download the lifecycle
manifest document.

The current status of `jsii` compiler release lines is also documented on the
[repository's `README.md` document][readme].

### Modification

The maintainers of the jsii project reserve the right to modify this Maintenance
and Support Policy as necessary. Updates will be proposed by way of a pull
request updating this document. Updates that materially affect the support
timeline of release lines will be broadly announced to the community via
established communication channels (such as the `cdk.dev` Slack), and will
remain open for the community to comment on for a minimum of 15 days.

Community members are welcome to propose changes to the support and maintenace
policy through the same process.

### Derogation

Under _exceptional_ circumstances, the project maintainers may elect to derogate
from this Support & Maintenance Policy. In cases where the decisision to
derogate extends supplemental maintenance & support coverage for a release line,

> the increased coverage will simply be documented in the

Reviewer: ```suggestion
the increased coverage will be documented in the
```

Reviewer: [non-blocking][nit] The adverb can be skipped with no semantic loss.

[repository's `README.md` document][readme].

However, in situations where the derogation results in reduced support for one
or more release lines, the decision to derogate is considered a "break glass"
event and maintainers will document the motivation for the derogation, as well

> as provide extensive post-mortem analysis to determine and address the reason

Reviewer: ```suggestion
as provide the post-mortem analysis to determine and address the reason
```

Reviewer: [non-blocking][nit] By Amazon standards any post-mortem should be extensive.
Since it's not immediately visible what level of detail is to be expected from the public's point of view, should the adjective be skipped?
